_posts/2022-12-19-running-outputs-outside-zonejs-for-angular-performance-optimization.md

```diff
@@ -0,0 +1,288 @@
+    // Get access to ngZone
+    const ngZone = inject(NgZone);
+
+    // Create a subject of the same generic type
+    // as the source observable
+    const sub$$ = new Subject<T>();
+    
+    // Declare the subscription here, we need to clean it up later
+    let subscription: Subscription;
+
+    // Run the actual subscription that executes an
+    // addEventListener outside of Angular, so that
+    // Change Detection won't be triggered automatically
+    ngZone.runOutsideAngular(() => {
+        subscription = source$.subscribe(sub$$);
+    });
+
+    // Return the subject but be sure to unsubscribe from
+    // the source observable when the subject completes
+    // or errors
+    return sub$$.pipe(
+        finalize(() => {
+            subscription.unsubscribe();
+        })
+    );
+};
+```
+
+Everything works as expected now: `ChildComponent` does not trigger Change Detection anymore but `ParentComponent` is notified with the new values.
+
+We can see that in this [StackBlitz example](https://stackblitz.com/edit/angular-ivy-komr7r){:target="_blank"}
+The DOM is not rerendered, but `console.log` statements are being made every second and when the user moves the mouse.
+
+## Custom Change Detection
+
+`ChildComponent` does what it needs to do:
+- Run event listeners in the outer zone
+- Notify the parent through `@Output()`s without triggering Change Detection
+
+`ParentComponent` is now responsible of triggering Change Detection and we don't want to trigger it all the time. We only want to trigger it when the mouse of the user moves within the green box.
+
+
+```typescript
+export class ParentComponent {
+    // Inject ChangeDetectorRef because we want to
+    // manually trigger Change Detection with the
+    // detectChanges() function
+    private readonly cdRef= inject(ChangeDetectorRef)
+    val = 0;
+    mouseX = null;
+    mouseY = null;
+
+    public onValChange(v: number): void {
+        // Set the value of val, but don't trigger
+        // Change Detection
+        this.val = v;
+        console.log(v);
+    }
+
+    public onMove(e: { x: number; y: number }): void {
+        // Set the values always
+        this.mouseX = e.x;
+        this.mouseY = e.y;
+        // Only trigger Change Detection when the event
+        // occured inside the green box
+        if(e.x <= 100 && e.y <= 100){
+            this.cdRef.detectChanges();
+        }
+        console.log(e);
+    }
+}
+```
+
+We can find the optimized Stackblitz example [here](https://stackblitz.com/edit/angular-ivy-zkbwrw){:target="_blank"}
+
+## Conclusion
+
+We learned that there are 3 ways of optimizing for Change Detection.
+The first way is by [using the OnPush Change Detection strategy which is described in detail in this article](https://blog.simplified.courses/angular-change-detection-onpush-or-not/){:target="_blank"}. The second one is detaching a component completely from the Change Detector and the third one is to run code in an outer zone.
```

Comment:
Make this a list as well for quicker scanning...

1. By [using the OnPush....] ...
2. Detaching a component ....
3. Run Code in an outer....